packages/salesforcedx-vscode-core/src/commands/commands.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {
+  CliCommandExecutor,
+  Command
+} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { channelService } from '../channels';
+import { notificationService } from '../notifications';
+import { isSfdxProjectOpened } from '../predicates';
+import { CancellableStatusBar, taskViewService } from '../statuses';
+
+// Precondition checking
+////////////////////////
+export interface PreconditionChecker {
+  check(): boolean;
+}
+
+export class SfdxWorkspaceChecker implements PreconditionChecker {
+  public check(): boolean {
+    const result = isSfdxProjectOpened.apply(vscode.workspace);
+    if (!result.result) {
+      notificationService.showErrorMessage(result.message);
+      return false;
+    }
+    return true;
+  }
+}
+
+// Input gathering
+//////////////////
+export interface ContinueResponse<T> {
+  type: 'CONTINUE';
+  data: T;
+}
+
+export interface CancelResponse {
+  type: 'CANCEL';
+}
+
+export interface ParametersGatherer<T> {
+  gather(): Promise<CancelResponse | ContinueResponse<T>>;
+}
+
+export class EmptyParametersGatherer implements ParametersGatherer<{}> {
```

Comment:
There are many commands that don't need additional parameters.
In that case, instead of special casing, we use an empty gatherer that always return true.
This similar to the [Null Object](https://en.wikipedia.org/wiki/Null_Object_pattern) pattern.